File: VFD.cpp
#include "Arduino.h"
#include "VFD.h"
/*
Create instance of display, arguments:
1) number of segment pins
2) array of segment pins, first 7 or 9 (see documentation) in conventional order (A-G) or (A-I) 
 for number segments if numbers wanted:
*   A
* F   B
*   G
* E   C
*   D 
*********
*   A
* F   B
*  GHI
* E   C
*   D 
*********
* With star-type digit (15 segment) you can use letter, message, and scrollingMessage
* functions, order segment pin array in this order:
*   A
* FJKLB
*  GHI
* EMNOC
*   D 
3) number of grid pins
4) array of grid pins, rightmost digit first
5) (optional) number of segments to be used for numbers (7 or 9)
*/
VFD::VFD(const byte numSegPins, const byte * segPins,const byte numGridPins, const byte * gridPins)
{
  _numSegPins = numSegPins;
  _segPins = segPins;
  _numGridPins = numGridPins;
  _gridPins = gridPins;
  _segsForNumbers = 7;
  for(byte i = 0; i<_numSegPins; i++) {
    pinMode(_segPins[i],OUTPUT);
  }
  for(byte i = 0; i<_numGridPins; i++) {
    pinMode(_gridPins[i],OUTPUT);
  }
  off(0);
}

VFD::VFD(const byte numSegPins, const byte * segPins,const byte numGridPins, const byte * gridPins, const byte segsForNumbers)
{
  _numSegPins = numSegPins;
  _segPins = segPins;
  _numGridPins = numGridPins;
  _gridPins = gridPins;
  _segsForNumbers = segsForNumbers;
  for(byte i = 0; i<_numSegPins; i++) {
    pinMode(_segPins[i],OUTPUT);
  }
  for(byte i = 0; i<_numGridPins; i++) {
    pinMode(_gridPins[i],OUTPUT);
  }
  off(0);
}

void VFD::on()
{
  for(byte k=0; k<_numGridPins; k++) {
    for (byte i = 0; i < _numSegPins; i++) {
      segment(k, i);
    }
  }
}

void VFD::off(int duration)
{
  for (byte i = 0; i < _numSegPins; i++) {
    digitalWrite(_segPins[i], HIGH);
  }
  for (byte i = 0; i < _numGridPins; i++) {
    digitalWrite(_gridPins[i], HIGH);
  } 
  delay(duration);
}

void VFD::blink(int duration) {
  if(isOff) {
    off(1);
  } else {
    on();
  }
  if(millis() > lastChange + duration) {
    isOff = !isOff;
    lastChange = millis();
  }
}

void VFD::segment(byte g, byte s) {
    digitalWrite(_gridPins[g], LOW);
    digitalWrite(_segPins[s], LOW);
}

void VFD::segment(byte g, byte s, int duration) {
  off(0);
  digitalWrite(_gridPins[g], LOW);
  digitalWrite(_segPins[s], LOW);
  delay(duration);
  digitalWrite(_gridPins[g], HIGH);
  digitalWrite(_segPins[s], HIGH);
}

void VFD::number(byte d, byte num) {
  for (byte i = 0; i < _segsForNumbers; i++) {
    if(_segsForNumbers == 9) {
      digitalWrite(_segPins[i], numbers9[num][i]);
    } else {
      digitalWrite(_segPins[i], numbers7[num][i]);
    }
  }
  for (byte j = 0; j < _numGridPins; j++) {
    if (j == d) {
      digitalWrite(_gridPins[j], LOW);
    } else {
      digitalWrite(_gridPins[j], HIGH);
    }
  }
  delay(5);
}

void VFD::multiDigitNumber(int num) {
  char onesChar, tensChar, hundredsChar, thousandsChar, tenThousandsChar;
  int ones, tens, hundreds, thousands, tenThousands;
  String numStr = String(num);

  if (num < 10) {
    onesChar = numStr[0];
    ones = byte(onesChar - '0');
    number(0, ones);
  } else if(num >=10 && num < 100) {
    onesChar = numStr[1];
    tensChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    number(0, ones);
    number(1, tens);
  } else if(num >=100 && num < 1000) {
    onesChar = numStr[2];
    tensChar = numStr[1];
    hundredsChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    hundreds = int(hundredsChar - '0');
    number(0, ones);
    number(1, tens);
    number(2, hundreds);
  } else if(num >= 1000 && num < 10000) {
    onesChar = numStr[3];
    tensChar = numStr[2];
    hundredsChar = numStr[1];
    thousandsChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    hundreds = int(hundredsChar - '0');
    thousands = int(thousandsChar - '0');
    number(0, ones);
    number(1, tens);
    number(2, hundreds);
    number(3, thousands);
  } else {
    onesChar = numStr[4];
    tensChar = numStr[3];
    hundredsChar = numStr[2];
    thousandsChar = numStr[1];
    tenThousandsChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    hundreds = int(hundredsChar - '0');
    thousands = int(thousandsChar - '0');
    tenThousands = int(tenThousandsChar - '0');
    number(0, ones);
    number(1, tens);
    number(2, hundreds);
    number(3, thousands);
    number(4, tenThousands);  
  }
}

void VFD::multiDigitNumber(byte g, int num) {
  char onesChar, tensChar, hundredsChar, thousandsChar, tenThousandsChar;
  int ones, tens, hundreds, thousands, tenThousands;
  String numStr = String(num);

  if (num < 10) {
    onesChar = numStr[0];
    ones = byte(onesChar - '0');
    number(g, ones);
  } else if(num >=10 && num < 100) {
    onesChar = numStr[1];
    tensChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    number(g, ones);
    number(g+1, tens);
  } else if(num >=100 && num < 1000) {
    onesChar = numStr[2];
    tensChar = numStr[1];
    hundredsChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    hundreds = int(hundredsChar - '0');
    number(g, ones);
    number(g+1, tens);
    number(g+2, hundreds);
  } else if(num >= 1000 && num < 10000) {
    onesChar = numStr[3];
    tensChar = numStr[2];
    hundredsChar = numStr[1];
    thousandsChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    hundreds = int(hundredsChar - '0');
    thousands = int(thousandsChar - '0');
    number(g, ones);
    number(g+1, tens);
    number(g+2, hundreds);
    number(g+3, thousands);
  } else {
    onesChar = numStr[4];
    tensChar = numStr[3];
    hundredsChar = numStr[2];
    thousandsChar = numStr[1];
    tenThousandsChar = numStr[0];
    ones = byte(onesChar - '0');
    tens = byte(tensChar - '0');
    hundreds = int(hundredsChar - '0');
    thousands = int(thousandsChar - '0');
    tenThousands = int(tenThousandsChar - '0');
    number(g, ones);
    number(g+1, tens);
    number(g+2, hundreds);
    number(g+3, thousands);
    number(g+4, tenThousands);  
  }
}

void VFD::crazyEights(int speed) {
  if(_segsForNumbers == 9) {
    byte segs[] = {0, 1, 8, 7, 6, 4, 3, 2, 8, 7, 6, 5};
    for (byte i = 0; i < sizeof(segs)/sizeof(segs[0]); i++) {
      for(byte j = 0; j < _numGridPins; j++) {
        segment(j, segs[i]);
      }
      delay(speed);
      off(1);   
    }   
  } else {
    byte segs[] = {0, 1, 6, 4, 3, 2, 6, 5};
    for (byte i = 0; i < sizeof(segs)/sizeof(segs[0]); i++) {
      for(byte j = 0; j < _numGridPins; j++) {
        segment(j, segs[i]);
      }
      delay(speed);
      off(1);   
    }
  }
}

void VFD::message(String text, byte firstGrid, byte lastGrid, int waitTime) {
  byte len = text.length();
  byte multiplier = lastGrid - firstGrid;
  if (millis() > lastTime + waitTime) {
    messageFragment++;
    lastTime = millis();
  }
  if (messageFragment * multiplier >= len) messageFragment = 0;
  messageSegment(text.substring(messageFragment * multiplier, messageFragment * multiplier + multiplier), firstGrid, lastGrid);
}

void VFD::scrollingMessage(String text, byte firstGrid, byte lastGrid, int waitTime) {
  String newText = "       " + text;
  byte len = newText.length();
  if (millis() > lastTime + waitTime / (lastGrid - firstGrid)) {
    messageFragment++;
    lastTime = millis();
  }
  if (messageFragment > len) messageFragment = 0;
  messageSegment(newText.substring(messageFragment, messageFragment + (lastGrid - firstGrid)), firstGrid, lastGrid);
}

void VFD::messageSegment(String mesg, byte firstGrid, byte lastGrid) {
	mesg.toLowerCase();
  for(byte i = firstGrid; i < lastGrid; i++) {
  	char let = mesg[i];
    letter(lastGrid - 1 - i, let);
  }
}

void VFD::letter(byte g, char whichLetter) {
  for (byte i = 0; i < _numGridPins; i++) {
    if (i == g) {
      digitalWrite(_gridPins[i], LOW);
    } else {
      digitalWrite(_gridPins[i], HIGH);
    }
  }
  byte tempLetter = ' ';
  switch (whichLetter) {
    case ' ':
      tempLetter = 0;
      break;
    case 'a':
      tempLetter = 1;
      break;
    case 'b':
      tempLetter = 2;
      break;
    case 'c':
      tempLetter = 3;
      break;
    case 'd':
      tempLetter = 4;
      break;
    case 'e':
      tempLetter = 5;
      break;
    case 'f':
      tempLetter = 6;
      break;
    case 'g':
      tempLetter = 7;
      break;
    case 'h':
      tempLetter = 8;
      break;
    case 'i':
      tempLetter = 9;
      break;
    case 'j':
      tempLetter = 10;
      break;
    case 'k':
      tempLetter = 11;
      break;
    case 'l':
      tempLetter = 12;
      break;
    case 'm':
      tempLetter = 13;
      break;
    case 'n':
      tempLetter = 14;
      break;
    case 'o':
      tempLetter = 15;
      break;
    case 'p':
      tempLetter = 16;
      break;
    case 'q':
      tempLetter = 17;
      break;
    case 'r':
      tempLetter = 18;
      break;
    case 's':
      tempLetter = 19;
      break;
    case 't':
      tempLetter = 20;
      break;
    case 'u':
      tempLetter = 21;
      break;
    case 'v':
      tempLetter = 22;
      break;
    case 'w':
      tempLetter = 23;
      break;
    case 'x':
      tempLetter = 24;
      break;
    case 'y':
      tempLetter = 25;
      break;
    case 'z':
      tempLetter = 26;
      break;
    case '0':
      tempLetter = 27;
      break;
    case '1':
      tempLetter = 28;
      break;
    case '2':
      tempLetter = 29;
      break;
    case '3':
      tempLetter = 30;
      break;
    case '4':
      tempLetter = 31;
      break;
    case '5':
      tempLetter = 32;
      break;
    case '6':
      tempLetter = 33;
      break;
    case '7':
      tempLetter = 34;
      break;
    case '8':
      tempLetter = 35;
      break;
    case '9':
      tempLetter = 36;
      break;
    default:
      tempLetter = 0;
      break;
  }
  for (byte i = 0; i < sizeof(alphanum[0]) / sizeof(alphanum[0][0]); i++) {
    digitalWrite(_segPins[i], alphanum[tempLetter][i]);
  }
  delay(3);
}

File: python/vfd.py
from gpiozero import LED
import time
class VFD:
    def __init__(self, grid_pins, seg_pins):
        self.grids = []
        self.segments = []
        self.is_off = False
        self.last_change = 0
        self.message_segment = 0
        self.max_message_length = 8 # set max length of display digits
        self.message_speed = 0.2
        self.scrolling = False
        '''
        Create instance of display, arguments:
        1) number of segment pins
        2) number of segment pins, first 7 in conventional order (A-G) 
        for number segments if numbers wanted:
        *   A
        * F   B
        *   G
        * E   C
        *   D 
        * With star-type digit (15 segment) you can use letter, message, and scrollingMessage
        * functions, order segment pin array in this order:
        *   A
        * FJKLB
        *  GHI
        * EMNOC
        *   D 
        '''
        self.numbers = [
            [0, 0, 0, 0, 0, 0, 1],#0    
            [1, 0, 0, 1, 1, 1, 1],#1
            [0, 0, 1, 0, 0, 1, 0],#2
            [0, 0, 0, 0, 1, 1, 0],#3
            [1, 0, 0, 1, 1, 0, 0],#4
            [0, 1, 0, 0, 1, 0, 0],#5
            [0, 1, 0, 0, 0, 0, 0],#6
            [0, 0, 0, 1, 1, 1, 1],#7
            [0, 0, 0, 0, 0, 0, 0],#8
            [0, 0, 0, 0, 1, 0, 0]#9
        ]
        self.alphanumeric = [
            [0, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1], #a
            [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],  #b
            [0, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1],  #c
            [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1],  #d
            [0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1],  #e
            [0, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1],  #f
            [0, 1, 0, 0, 0, 0, 1, 1, 0, 1, 1, 1, 1, 1, 1],  #g
            [1, 0, 0, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],  #h
            [0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1],  #i
            [1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],  #j
            [1, 1, 1, 1, 0, 0, 0, 0, 1, 1, 1, 0, 1, 1, 0],  #k
            [1, 1, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1],  #l
            [1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 0, 1, 1, 1],  #m
            [1, 0, 0, 1, 0, 0, 1, 0, 1, 0, 1, 1, 1, 1, 0],  #n
            [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1],  #o
            [0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],  #p
            [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 0, 1, 1],  #q
            [0, 0, 1, 1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 0],  #r
            [0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],  #s
            [0, 1, 1, 1, 1, 1, 1, 0, 1, 1, 0, 1, 1, 0, 1],  #t
            [1, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1],  #u
            [1, 1, 1, 1, 0, 0, 1, 0, 1, 1, 1, 0, 0, 1, 1],  #v
            [1, 0, 0, 1, 0, 0, 1, 0, 1, 1, 1, 1, 0, 1, 0],  #w
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 0, 1, 0],  #x
            [1, 1, 1, 1, 1, 1, 1, 0, 1, 0, 1, 0, 1, 0, 1],  #y
            [0, 1, 1, 0, 1, 1, 1, 0, 1, 1, 1, 0, 0, 1, 1],  #z
            [0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1], #0
            [1, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],#1
            [0, 0, 1, 0, 0, 1, 0, 0, 0, 1, 1, 1, 1, 1, 1],#2
            [0, 0, 0, 0, 1, 1, 1, 0, 0, 1, 1, 1, 1, 1, 1],#3
            [1, 0, 0, 1, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],#4
            [0, 1, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],#5
            [0, 1, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],#6
            [0, 0, 0, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1],#7
            [0, 0, 0, 0, 0, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],#8
            [0, 0, 0, 0, 1, 0, 0, 0, 0, 1, 1, 1, 1, 1, 1],#9
            [1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1, 1] #space
        ]
        # assign LED object pins
        for grid in grid_pins:
            self.grids.append(LED(grid))
        for segment in seg_pins:
            self.segments.append(LED(segment))
        
        self.set_off()

    def get_scrolling(self):
        return self.scrolling

    def segment(self, grid, seg, duration=None):
        if duration == None:
            self.grids[grid].off()
            self.segments[seg].off()
        else:
            self.set_off()
            self.grids[grid].off()
            self.segments[seg].off()
            time.sleep(duration)
            self.grids[grid].on()
            self.segments[seg].on()

    def set_on(self, duration=None):
        if duration is None:
            for grid in self.grids:
                grid.off()
            for seg in self.segments:
                seg.off()
        else:
            for grid in self.grids:
                grid.off()
            for seg in self.segments:
                seg.off()
            time.sleep(duration)
            for grid in self.grids:
                grid.on()
            for seg in self.segments:
                seg.on()

    def set_off(self, duration=None):
        if duration is None:
            for grid in self.grids:
                grid.on()
            for seg in self.segments:
                seg.on()
        else:
            for grid in self.grids:
                grid.on()
            for seg in self.segments:
                seg.on()
            time.sleep(duration)
            for grid in self.grids:
                grid.off()
            for seg in self.segments:
                seg.off()

    def blink(self, speed, duration):
        start_time = time.time()
        self.last_change = time.time()
        self.scrolling = True
        while self.scrolling:
            if self.is_off:
                self.set_off()
            else:
                self.set_on()
            if time.time() > self.last_change:
                self.is_off = not self.is_off
                self.last_change = time.time() + speed
            if time.time() > start_time + duration:
                self.scrolling = False

    def number(self, grid, number):
        self.set_off()
        self.grids[grid].off()
        for (val,seg) in enumerate(self.numbers[number]):
            if seg == 0:
                self.segments[val].off()
            else:
                self.segments[val].on()
        time.sleep(0.005)

    def multi_digit_number(self, number):
        if number > 10000: number = 9999
        if number < 0: number = 0
        split_num = str(number) #split digits into str to get separate elemts
        self.set_off()
        if len(split_num) == 1:
            self.number(0, number)
        elif len(split_num) == 2:
            self.number(0, int(split_num[1]))
            self.number(1, int(split_num[0]))
        elif len(split_num) == 3:
            self.number(0, int(split_num[2]))
            self.number(1, int(split_num[1]))
            self.number(2, int(split_num[0]))
        elif len(split_num) == 4:
            self.number(0, int(split_num[3]))
            self.number(1, int(split_num[2]))
            self.number(2, int(split_num[1]))
            self.number(3, int(split_num[0]))

    def message(self, text):
        text = text.lower()
        for i in range(len(text)):
            self.letter(i, text[len(text)-i-1])

    def scrolling_message(self, text):
        max_length = self.max_message_length
        buffer = ""
        for i in range(max_length):
            buffer += " "
        self.scrolling = True
        while self.scrolling:
            new_text = buffer + text + buffer
            msg_length = len(new_text)
            if time.time() > self.last_change:
                self.message_segment += 1
                self.last_change = time.time() + self.message_speed
            if self.message_segment + max_length >= msg_length:
                self.scrolling = False
                self.message_segment = 0
            self.send_message_segment(new_text[self.message_segment:self.message_segment + max_length], max_length)

    def send_message_segment(self, message, max_length):
        message = message.lower()
        for i in range(max_length):
            self.letter(i,message[max_length - 1 - i])

    def letter(self, grid, which_letter):
        self.set_off()
        for i in range(len(self.grids)):
            if i == grid:
                self.grids[i].off()
            else:
                self.grids[i].on()
        alpha_to_num = {
            'a': 0,
            'b': 1,
            'c': 2,
            'd': 3,
            'e': 4,
            'f': 5,
            'g': 6,
            'h': 7,
            'i': 8,
            'j': 9,
            'k': 10,
            'l': 11,
            'm': 12,
            'n': 13,
            'o': 14,
            'p': 15,
            'q': 16,
            'r': 17,
            's': 18,
            't': 19,
            'u': 20,
            'v': 21,
            'w': 22,
            'x': 23,
            'y': 24,
            'z': 25,
            '0': 26,
            '1': 27,
            '2': 28,
            '3': 29,
            '4': 30,
            '5': 31,
            '6': 32,
            '7': 33,
            '8': 34,
            '9': 35,
            ' ': 36
        }
        for key, value in alpha_to_num.items():
            if which_letter == key:
                for (val,seg) in enumerate(self.alphanumeric[value]):
                    if seg == 0:
                        self.segments[val].off()
                    else:
                        self.segments[val].on()
        time.sleep(0.005)
